src/Common/src/TypeSystem/IL/Stubs/StructMarshallingThunk.cs

```diff
@@ -0,0 +1,253 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Interop;
+using System.Collections.Generic;
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.IL.Stubs
+{
+    internal enum StructMarshallingThunkType : byte
+    {
+        MangedToNative = 1,
+        NativeToManage = 2,
+        Cleanup = 4
+    };
+
+    internal class StructMarshallingThunk : ILStubMethod
+    {
+        internal readonly TypeDesc ManagedType;
+        internal readonly NativeStructType NativeType;
+        internal readonly StructMarshallingThunkType ThunkType;
+        private  InteropStateManager _interopStateManager;
+        private TypeDesc _owningType;
+
+        public StructMarshallingThunk(TypeDesc owningType, TypeDesc managedType, StructMarshallingThunkType thunkType, InteropStateManager interopStateManager)
+        {
+            _owningType = owningType;
+            ManagedType = managedType;
+            _interopStateManager = interopStateManager;
+            NativeType = _interopStateManager.GetStructMarshallingNativeType((MetadataType)managedType);
```

Comment:
Nit: if this code is only able to operate if `managedType` is `MetadataType`, it should just declare it in the signature.